### ISSUE TITLE
Optional replacement of input buffer with a callback

### ISSUE THREAD

**Issue by incrediball** [lieff/minimp3#44]
Like hdt551, I'm hoping to use the decoder on an embedded microcontroller (ARM Cortex M4) but the current memory usage is an issue. The required large size of the input buffer is something that could possible be eliminated if an optional method that uses a callback function could be implemented. I used this method for another audio decoder and it was called from the get_bits function when more data were needed. The callback function returned a pointer to a smaller buffer along with the number of bytes loaded. These data were then consumed by get_bits until empty and then the callback was called again. This approach would avoid having to load up to the recommend 16K for 10 frames since the data are always on demand.

I did have a look at the source code and I saw a number of problems since get_bits does not appear to be the only thing accessing the data. However I'm not sure how some of it works (such as the inner workings of L3_huffman how all this works with the maindata array) so I don't know if it's easily possible.

Further thoughts:
*If this approach could also enable the removal of the maindata array then that would also be a big win.
**The "callback" function need not be a function pointer, it could simply be a function declared by minimp3.h and the C code that includes the minimp3.h file with the MINIMP3_IMPLEMENTATION definition would implement it.
***If necessary, the function could be allowed to seek forward (or backward if it helps to make the bit reservoir unnecessary) for accessing required data.

**Comment by incrediball** [comment]
Thanks for the information. Regarding the "sync procedure", if the mp3 file is played from the beginning to the end without skipping forward or back, is any sync procedure required? I am not planning on providing skip forward or back functions so can I just reduce MAX_FRAME_SYNC_MATCHES down to 1 and use the smaller buffer size without worry that it will have problems processing some files?

**Comment by lieff** [comment]
Yes, it's required. There valid mp3 headers can be found in middle of the frame, and for vbr case, decoder do not knows next frame size + there can be padding. Also, well, usually there no garbage between frames, I've not seen such files in the wild, but reference vectors have such case, so it still considered as valid file.
I would recommend MAX_FRAME_SYNC_MATCHES at least 2-3 and at least ~MAX_L3_FRAME_PAYLOAD_BYTES*2 buffer size. Also I recommend mmap 4096-8192 bytes (1-2 pages) twice as hardware ring buffer, to not make memcpy() during keeping input buffer MAX_L3_FRAME_PAYLOAD_BYTES*2 size. But not all microcontrollers supports that.
